src/arrow_reader/column/binary.rs

```diff
@@ -49,20 +40,23 @@ pub fn new_binary_iterator(
 
 pub struct DirectBinaryIterator {
     values: Box<Values<Decompressor>>,
-    lengths: Box<dyn Iterator<Item = error::Result<u64>> + Send>,
+    lengths: Box<dyn Iterator<Item = Result<u64>> + Send>,
+}
+
+impl DirectBinaryIterator {
+    fn iter_next(&mut self) -> Result<Option<Vec<u8>>> {
+        let next = match self.lengths.next() {
+            Some(length) => Some(self.values.next(length? as usize)?.to_vec()),
+            None => None,
+        };
+        Ok(next)
+    }
 }
 
 impl Iterator for DirectBinaryIterator {
-    type Item = error::Result<Vec<u8>>;
+    type Item = Result<Vec<u8>>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        match self.lengths.next() {
-            Some(Ok(length)) => match self.values.next(length as usize) {
-                Ok(value) => Some(Ok(value.to_vec())),
-                Err(err) => Some(Err(err)),
-            },
-            Some(Err(err)) => Some(Err(err)),
-            None => None,
-        }
+        self.iter_next().transpose()
```

Comment:
Idea of this pattern is to be able to utilize `?` to simplify code a bit (no need to handle `Some(Ok(length))` which then has to again match)